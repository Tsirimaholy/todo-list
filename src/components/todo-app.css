.container{
    display: flex;
}

.todo {
    border: 1px solid black;
    width: 100%;
    margin: 2em;
    padding-left: 2em;
    padding-right: 2em;
    padding-bottom: 5em;
    border-radius: 1em;
}