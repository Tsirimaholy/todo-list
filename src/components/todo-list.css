.task_list__container{
    height: 30vh;
    border-radius: 2%;
}

.task_list{
    overflow-y: scroll;
    height: inherit;
}

.todo_list__form {
    width: 100%;
}
.todo_list__item {
    list-style-type: none;
}