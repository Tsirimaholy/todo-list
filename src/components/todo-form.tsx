import React, {FC, useRef, useState} from "react";
import {TTask} from "./todo-list";

type TTodoForm = {
    onSubmit: (task: TTask) => void;
    defaultValue?: string;
    visible?: boolean
}
export const TodoForm: FC<TTodoForm> = (props) => {
    const {onSubmit, defaultValue, visible = true} = props;
    const [currentTask, setCurrentTask] = useState<TTask>({name: defaultValue || "", done: false});
    const handleAdd = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if ((e.key !== 'Enter' && currentTask.name.length > 0) || currentTask.name.length <= 0) return
        onSubmit(currentTask);
        // reset the current task value
        setCurrentTask({name: "", done: false})
    }

    function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
        setCurrentTask({...currentTask, name: e.target.value})
    }

    if (!visible) return null;
    return (
        <div className="todo_list__form">
            <input
                type="text"
                placeholder="todo-input"
                onKeyDown={handleAdd}
                onChange={handleChange}
                value={currentTask.name}
            />
        </div>
    );
};