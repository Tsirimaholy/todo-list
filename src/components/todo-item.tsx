import React from "react";
import {TTask} from "./todo-list";

type TodoItemProps = { isDone?: boolean, onClick: () => void, task: TTask };
export const TodoItem = (props: TodoItemProps) => {
    const {isDone = false, onClick, task} = props;


    return (<>
        {!isDone && <input type={"checkbox"} onClick={onClick}/>}
        <span>{task.name}</span>
    </>)
};