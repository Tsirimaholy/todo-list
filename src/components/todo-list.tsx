import React, {FC} from 'react';
import {TodoForm} from "./todo-form";
import "./todo-list.css"
import {TodoItem} from "./todo-item";

export type TTask = {
    name: string;
    done: boolean
}


type TTodoList = {
    taskList: TTask[];
    hideTodoForm?: boolean
    onInsert?: (task: TTask) => void
    onDone?: (task: TTask) => void
}

const TodoList: FC<TTodoList> = ({taskList, onInsert, onDone, hideTodoForm = false}) => {
    return (
        <div className="task_list__container">
            <TodoForm onSubmit={task => onInsert && onInsert(task)} visible={!hideTodoForm}/>
            <ul className="task_list">
                {taskList.map((task, index) =>
                    task.name && (
                        <li className="todo_list__item" key={index}>
                            <TodoItem onClick={() => onDone && onDone(task)} task={task}/>
                        </li>
                    ))}
            </ul>
        </div>
    );
};


export default TodoList;