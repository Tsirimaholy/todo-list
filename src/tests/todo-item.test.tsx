import {create} from "react-test-renderer";
import {TodoItem} from "../components/todo-item";
import React from "react";
import {TTask} from "../components/todo-list";
import {cleanup, render, screen} from "@testing-library/react";

afterEach(() => {
    cleanup();
})


describe('Todo Item', () => {
    it('should renders correctly', function () {
        const onCLickMock = jest.fn();
        const taskMock: TTask = {
            name: "some task",
            done: false
        };

        const renderer = create(<TodoItem onClick={onCLickMock} task={taskMock}/>);
        const toJSON = renderer.toJSON();
        expect(toJSON).toMatchSnapshot();
    });


    it('should display the provided task name', function () {
        const onCLickMock = jest.fn();
        const taskMock: TTask = {
            name: "some task",
            done: false
        };

        render(<TodoItem onClick={onCLickMock} task={taskMock}/>);

        const text = screen.getByText("some task");
        expect(text).toBeInTheDocument();
    });

    it('should display display a checkbox of a task that was not done yet', () => {
        const onClickMock = jest.fn();
        const taskMock: TTask = {
            name: "some task",
            done: false
        };

        render(<TodoItem onClick={onClickMock} task={taskMock}/>);

        const checkBox = screen.getByRole("checkbox") as HTMLInputElement;
        expect(checkBox).toBeInTheDocument();
        // the checkbox is not checked
        expect(checkBox.checked).toBeFalsy();
    });

    it('should "not" display display a checkbox on a done task', () => {
        const onClickMock = jest.fn();
        const taskMock: TTask = {
            name: "some task",
            done: true
        };

        render(<TodoItem onClick={onClickMock} task={taskMock}/>);

        const {length} = screen.queryAllByRole("checkbox");
        expect(length).toEqual(0);

        const textWrapper = screen.getByText("some task") as HTMLSpanElement;
        expect(textWrapper).toBeTruthy();
    });
});