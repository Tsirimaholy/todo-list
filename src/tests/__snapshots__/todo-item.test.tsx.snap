// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Item should renders correctly 1`] = `
Array [
  <input
    onClick={[MockFunction]}
    type="checkbox"
  />,
  <span>
    some task
  </span>,
]
`;
